core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```diff
@@ -0,0 +1,387 @@
+  }
+
+  case class WitnessUTXO(spentWitnessTransaction: TransactionOutput)
+      extends InputPSBTRecord {
+    override val key: ByteVector = hex"01"
+    override val value: ByteVector = spentWitnessTransaction.bytes
+  }
+
+  case class PartialSignature(
+      pubKey: ECPublicKey,
+      signature: ECDigitalSignature)
+      extends InputPSBTRecord {
+    require(pubKey.size == 33, s"pubKey must be 33 bytes, got: ${pubKey.size}")
+
+    override val key: ByteVector = hex"02" ++ pubKey.bytes
+    override val value: ByteVector = signature.bytes
+  }
+
+  case class SigHashType(hashType: HashType) extends InputPSBTRecord {
+    override val key: ByteVector = hex"03"
+    override val value: ByteVector = hashType.num.bytes
+  }
+
+  case class RedeemScript(redeemScript: ScriptPubKey) extends InputPSBTRecord {
+    override val key: ByteVector = hex"04"
+    override val value: ByteVector = redeemScript.asmBytes
+  }
+
+  case class WitnessScript(witnessScript: ScriptPubKey)
+      extends InputPSBTRecord {
+    override val key: ByteVector = hex"05"
+    override val value: ByteVector = witnessScript.asmBytes
+  }
+
+  case class BIP32DerivationPath(
+      pubKey: ECPublicKey,
+      masterFingerprint: ByteVector,
+      path: BIP32Path)
+      extends InputPSBTRecord {
+    require(pubKey.size == 33, s"pubKey must be 33 bytes, got: ${pubKey.size}")
+
+    override val key: ByteVector = hex"06" ++ pubKey.bytes
+    override val value: ByteVector =
+      path.path.foldLeft(masterFingerprint)(_ ++ _.toUInt32.bytes)
+  }
+
+  case class FinalizedScriptSig(scriptSig: ScriptSignature, extra: ByteVector)
+      extends InputPSBTRecord {
+    override val key: ByteVector = hex"07"
+    override val value: ByteVector = scriptSig.bytes ++ extra
+  }
+
+  case class FinalizedScriptWitness(scriptWitness: ScriptWitness)
+      extends InputPSBTRecord {
+    override val key: ByteVector = hex"08"
+    override val value: ByteVector = scriptWitness.bytes
+  }
+
+  case class ProofOfReservesCommitment(porCommitment: ByteVector)
+      extends InputPSBTRecord {
+    override val key: ByteVector = hex"09"
+    override val value: ByteVector = porCommitment
+  }
+
+  case class Unknown(key: ByteVector, value: ByteVector) extends InputPSBTRecord
+}
+
+sealed trait OutputPSBTRecord extends PSBTRecord
+
+object OutputPSBTRecord {
+
+  case class RedeemScript(redeemScript: ScriptPubKey) extends OutputPSBTRecord {
+    override val key: ByteVector = hex"00"
+    override val value: ByteVector = redeemScript.asmBytes
+  }
+
+  case class WitnessScript(witnessScript: ScriptPubKey)
+      extends OutputPSBTRecord {
+
+    override val key: ByteVector = hex"01"
+    override val value: ByteVector = witnessScript.asmBytes
+  }
+
+  case class BIP32DerivationPath(
+      pubKey: ECPublicKey,
+      masterFingerprint: ByteVector,
+      path: BIP32Path)
+      extends OutputPSBTRecord {
+    require(pubKey.size == 33, s"pubKey must be 33 bytes, got: ${pubKey.size}")
```

Comment:
Should this be moved to `ECPublickKey`?

Reply:
No, so it turns out there are a couple different serializations of `ECPublicKey` which are valid in different contexts. We want the compressed version here (hence we require 33 bytes)